clouddrift/kinematics.py

```diff
@@ -667,3 +724,147 @@ def velocity_from_position(
+
+        1. "forward" (default): finite difference is evaluated as ``dx[i] = dx[i+1] - dx[i]``;
+        2. "backward": finite difference is evaluated as ``dx[i] = dx[i] - dx[i-1]``;
+        3. "centered": finite difference is evaluated as ``dx[i] = (dx[i+1] - dx[i-1]) / 2``.
+
+    Forward and backward schemes are effectively the same except that the
+    position at which the velocity is evaluated is shifted one element down in
+    the backward scheme relative to the forward scheme. In the case of a
+    forward or backward difference scheme, the last or first element of the
+    velocity, respectively, is extrapolated from its neighboring point. In the
+    case of a centered difference scheme, the start and end boundary points are
+    evaluated using the forward and backward difference scheme, respectively.
+
+    Reference for the definition of spin is:
+
+    Veneziani, M., Griffa, A., Reynolds, A.M., Garraffo, Z.D. and Chassignet,
+    E.P., 2005. Parameterizations of Lagrangian spin statistics and particle
+    dispersion in the presence of coherent vortices. Journal of Marine Research,
+    63(6), pp.1057-1083.
+
+    Parameters
+    ----------
+    u : np.ndarray
+        Zonal velocity
+    v : np.ndarray
+        Meridional velocity
+    time : array-like
+        Time
+    difference_scheme : str, optional
+        Difference scheme to use; possible values are "forward", "backward", and "centered".
+    time_axis : int, optional
+        Axis along which the time varies (default is -1)
+
+    Returns
+    -------
+    s : float or np.ndarray
+        Spin
+
+    Raises
+    ------
+    ValueError
+        If u and v do not have the same shape.
+        If the time axis is outside of the valid range ([-1, N-1]).
+        If lengths of u, v, and time along time_axis are not equal.
+        If difference_scheme is not "forward", "backward", or "centered".
+
+    Examples
+    --------
+    TODO
+    """
+    if not u.shape == v.shape:
+        raise ValueError("u and v arrays must have the same shape.")
+
+    if not time.shape == u.shape:
+        if not time.size == u.shape[time_axis]:
+            raise ValueError("time must have the same length as u along time_axis.")
+
+    # axis must be in valid range
+    if time_axis < -1 or time_axis > len(u.shape) - 1:
+        raise ValueError(
+            f"axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(u.shape) - 1}])."
+        )
+
+    # Swap axes so that we can differentiate along the last axis.
+    # This is a syntax convenience rather than memory access optimization:
+    # np.swapaxes returns a view of the array, not a copy, if the input is a
+    # NumPy array. Otherwise, it returns a copy.
+    u = np.swapaxes(u, time_axis, -1)
+    v = np.swapaxes(v, time_axis, -1)
+    time = np.swapaxes(time, time_axis, -1)
+
+    if not time.shape == u.shape:
+        # time is 1-d array; broadcast to u.shape.
+        time = np.broadcast_to(time, u.shape)
+
```

Comment:
Make this an option with a keyword argument set to False by default such as `anomaly=False` i.e. by default the input velocities are not anomalies. I am thinking of an implementation for which a mean spatial field is removed first from the velocities. In that case the user passes `anomaly = True`.

Reply:
Does `anomaly=False` mean that we compute KE and spin based on full velocities (as input)? Likewise, if `anomaly=True`, we compute EKE by demeaning the input velocities first (current implementation)?

Or, do you mean that `anomaly=False` is a way for the user to communicate whether the input velocity is a full velocity or an anomaly and we demean or not based on that information?

Reply:
`anomaly=False` means the input velocities are not anomalies and hence we estimate the mean first as the trajectory mean in order to calculate the velocity anomalies and thus the spin and the eke.

Reply:
OK. And what do we do when `anomaly=True`?

Reply:
`anomalies=True` means the input velocity components are already the u' and v' of your formula. 

Which makes me think that the `eddy_kinetic_energy` function should probably be replaced by a `kinetic_energy` function which calculates `np.mean(u**2) + np.mean(v**2)`. If the input is an anomaly it is an EKE, if the input is not an anomaly it is a total kinetic energy.

Reply:
That's an option which I support. In this case the code should not remove the Lagrangian mean from the input.